<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Analysis &mdash; uga 2014.01 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2014.01',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="uga 2014.01 documentation" href="index.html" />
    <link rel="next" title="Meta-Analysis" href="meta-analysis.html" />
    <link rel="prev" title="Universal Genome Analyst" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="meta-analysis.html" title="Meta-Analysis"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Universal Genome Analyst"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">uga 2014.01 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="analysis">
<h1>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">Â¶</a></h1>
<pre class="literal-block">
usage: uga analyze [-h] [--version] [-o] [-q QSUB] [--name NAME]
                   [-d DIRECTORY] [--cpus CPUS] [--mem MEM]
                   [-r REGION | --region-list REGION_LIST]
                   [-s | -n SPLIT_N | --chr CHR | --split-chr]
                   [-j JOB | --job-list JOB_LIST] --data DATA --out OUT
                   --samples SAMPLES --pheno PHENO --model MODEL --fid FID
                   --iid IID --method
                   {gee_gaussian,gee_binomial,glm_gaussian,glm_binomial,lme_gaussian,lme_binomial,coxph,efftests}
                   [--focus FOCUS] [--sig SIG] [--sex SEX] [--male MALE]
                   [--female FEMALE] [--buffer BUFFER] [--miss MISS]
                   [--freq FREQ] [--rsq RSQ] [--hwe HWE] [--case CASE]
                   [--ctrl CTRL] [--nofail]
</pre>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-o=False</span>, <span class="option">--overwrite=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>overwrite existing output files</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-q</span>, <span class="option">--qsub</span></kbd></td>
<td>a group ID under which to submit jobs to the queue</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--name</span></kbd></td>
<td>a job name (only used with &#8211;qsub; if not set, &#8211;out basename will be used</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-d</span>, <span class="option">--directory</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>an output directory path</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--cpus=1</span></kbd></td>
<td>number of cpus (limited module availability)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--mem=3</span></kbd></td>
<td>amount of ram memory to request for queued job (in gigabytes)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-r</span>, <span class="option">--region</span></kbd></td>
<td>a region specified in tabix format (ie. 1:10583-1010582).</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--region-list</span></kbd></td>
<td>a filename for a list of tabix format regions</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-s=False</span>, <span class="option">--split=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>split region list entirely into separate jobs (requires &#8211;region-list)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-n</span>, <span class="option">--split-n</span></kbd></td>
<td>split region list into n separate jobs (requires &#8211;region-list)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--chr</span></kbd></td>
<td>chromosome number (1-26)</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--split-chr=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>split by chromosome (will generate up to 26 separate jobs depending on chromosome coverage)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-j</span>, <span class="option">--job</span></kbd></td>
<td>run a particular job number (requires &#8211;split-n)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--job-list</span></kbd></td>
<td>a filename for a list of job numbers (requires &#8211;split-n)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--data</span></kbd></td>
<td>a genomic data file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--out</span></kbd></td>
<td>an output file name (basename only: do not include path)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--samples</span></kbd></td>
<td>a sample file (single column list of IDs in order of data)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--pheno</span></kbd></td>
<td>a tab delimited phenotype file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--model</span></kbd></td>
<td>a comma separated list of models in the format &#8220;phenotype~age+factor(sex)+pc1+pc2+pc3+marker&#8221;</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fid</span></kbd></td>
<td>the column name with family ID</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--iid</span></kbd></td>
<td>the column name with sample ID</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--method</span></kbd></td>
<td><p class="first">the analysis method</p>
<p class="last">Possible choices: gee_gaussian, gee_binomial, glm_gaussian, glm_binomial, lme_gaussian, lme_binomial, coxph, efftests</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--focus</span></kbd></td>
<td>a comma separated list of variables for which stats will be output (default: marker)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--sig=5</span></kbd></td>
<td>significant digits to include in output (default: 5)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--sex</span></kbd></td>
<td>name of the column containing male/female status (requires MALE_CODE and female)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--male=1</span></kbd></td>
<td>the code for a male in sex (requires &#8211;sex and &#8211;female)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--female=2</span></kbd></td>
<td>the code for a male in sex (requires &#8211;sex and &#8211;female)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--buffer=100</span></kbd></td>
<td>a value for number of markers calculated at a time (WARNING: this argument will affect RAM memory usage; default: 100)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--miss</span></kbd></td>
<td>a threshold value for missingness</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--freq</span></kbd></td>
<td>a threshold value for allele frequency</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--rsq</span></kbd></td>
<td>a threshold value for r-squared (imputation quality)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--hwe</span></kbd></td>
<td>a threshold value for Hardy Weinberg p-value</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--case=1</span></kbd></td>
<td>the code for a case in the dependent variable column (requires ctrl; binomial fxn family only; default: 1)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ctrl=0</span></kbd></td>
<td>the code for a control in the dependent variable column (requires case; binomial fxn family only; default: 0)</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--nofail=False</span></kbd></td>
<td>exclude filtered/failed analyses from results</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Universal Genome Analyst</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="meta-analysis.html"
                        title="next chapter">Meta-Analysis</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/analysis.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="meta-analysis.html" title="Meta-Analysis"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Universal Genome Analyst"
             >previous</a> |</li>
        <li><a href="index.html">uga 2014.01 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Ryan Koesterer.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>