% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{uga Documentation}
\date{November 28, 2014}
\release{2014.01}
\author{Ryan Koesterer}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@mb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\renewcommand\PYGZsq{\textquotesingle}

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Universal Genome Analyst (uga) is a pipeline for the distribution and
management of large and small scale analysis of genetic data.


\chapter{Modules}
\label{index:universal-genome-analyst}\label{index:modules}

\section{Analysis}
\label{analyze::doc}\label{analyze:analysis}\begin{alltt}
usage: uga analyze {[}-h{]} {[}--version{]} {[}-o{]} {[}-q QSUB{]} {[}--name NAME{]}
                   {[}-d DIRECTORY{]} {[}--cpus CPUS{]} {[}--mem MEM{]}
                   {[}-r REGION \textbar{} --region-list REGION\_LIST{]}
                   {[}-s \textbar{} -n SPLIT\_N \textbar{} --chr CHR \textbar{} --split-chr{]}
                   {[}-j JOB \textbar{} --job-list JOB\_LIST{]} --data DATA --out OUT
                   --samples SAMPLES --pheno PHENO --model MODEL --fid FID
                   --iid IID --method
                   \{gee\_gaussian,gee\_binomial,glm\_gaussian,glm\_binomial,lme\_gaussian,lme\_binomial,coxph,efftests\}
                   {[}--focus FOCUS{]} {[}--sig SIG{]} {[}--sex SEX{]} {[}--male MALE{]}
                   {[}--female FEMALE{]} {[}--buffer BUFFER{]} {[}--miss MISS{]}
                   {[}--freq FREQ{]} {[}--rsq RSQ{]} {[}--hwe HWE{]} {[}--case CASE{]}
                   {[}--ctrl CTRL{]} {[}--nofail{]}
\end{alltt}
\begin{description}
\item[{Options:}] \leavevmode\begin{optionlist}{3cm}
\item [-{-}version]  
show program's version number and exit
\item [-o=False, -{-}overwrite=False]  
overwrite existing output files
\item [-q, -{-}qsub]  
a group ID under which to submit jobs to the queue
\item [-{-}name]  
a job name (only used with --qsub; if not set, --out basename will be used
\item [-d=/usr3/bustaff/koesterr/UGA/doc, -{-}directory=/usr3/bustaff/koesterr/UGA/doc]  
an output directory path
\item [-{-}cpus=1]  
number of cpus (limited module availability)
\item [-{-}mem=3]  
amount of ram memory to request for queued job (in gigabytes)
\item [-r, -{-}region]  
a region specified in tabix format (ie. 1:10583-1010582).
\item [-{-}region-list]  
a filename for a list of tabix format regions
\item [-s=False, -{-}split=False]  
split region list entirely into separate jobs (requires --region-list)
\item [-n, -{-}split-n]  
split region list into n separate jobs (requires --region-list)
\item [-{-}chr]  
chromosome number (1-26)
\item [-{-}split-chr=False]  
split by chromosome (will generate up to 26 separate jobs depending on chromosome coverage)
\item [-j, -{-}job]  
run a particular job number (requires --split-n)
\item [-{-}job-list]  
a filename for a list of job numbers (requires --split-n)
\item [-{-}data]  
a genomic data file
\item [-{-}out]  
an output file name (basename only: do not include path)
\item [-{-}samples]  
a sample file (single column list of IDs in order of data)
\item [-{-}pheno]  
a tab delimited phenotype file
\item [-{-}model]  
a comma separated list of models in the format ``phenotype\textasciitilde{}age+factor(sex)+pc1+pc2+pc3+marker''
\item [-{-}fid]  
the column name with family ID
\item [-{-}iid]  
the column name with sample ID
\item [-{-}method]  
the analysis method

Possible choices: gee\_gaussian, gee\_binomial, glm\_gaussian, glm\_binomial, lme\_gaussian, lme\_binomial, coxph, efftests
\item [-{-}focus]  
a comma separated list of variables for which stats will be output (default: marker)
\item [-{-}sig=5]  
significant digits to include in output (default: 5)
\item [-{-}sex]  
name of the column containing male/female status (requires MALE\_CODE and female)
\item [-{-}male=1]  
the code for a male in sex (requires --sex and --female)
\item [-{-}female=2]  
the code for a male in sex (requires --sex and --female)
\item [-{-}buffer=100]  
a value for number of markers calculated at a time (WARNING: this argument will affect RAM memory usage; default: 100)
\item [-{-}miss]  
a threshold value for missingness
\item [-{-}freq]  
a threshold value for allele frequency
\item [-{-}rsq]  
a threshold value for r-squared (imputation quality)
\item [-{-}hwe]  
a threshold value for Hardy Weinberg p-value
\item [-{-}case=1]  
the code for a case in the dependent variable column (requires ctrl; binomial fxn family only; default: 1)
\item [-{-}ctrl=0]  
the code for a control in the dependent variable column (requires case; binomial fxn family only; default: 0)
\item [-{-}nofail=False]  
exclude filtered/failed analyses from results
\end{optionlist}

\end{description}


\section{Meta-Analysis}
\label{meta:meta-analysis}\label{meta::doc}\begin{alltt}
usage: uga meta {[}-h{]} {[}--version{]} {[}-o{]} {[}-q QSUB{]} {[}--name NAME{]} {[}-d DIRECTORY{]}
                {[}--cpus CPUS{]} {[}--mem MEM{]}
                {[}-r REGION \textbar{} --region-list REGION\_LIST{]}
                {[}-s \textbar{} -n SPLIT\_N \textbar{} --chr CHR \textbar{} --split-chr{]}
                {[}-j JOB \textbar{} --job-list JOB\_LIST{]} -c CFG {[}-v VARS{]}
\end{alltt}
\begin{description}
\item[{Options:}] \leavevmode\begin{optionlist}{3cm}
\item [-{-}version]  
show program's version number and exit
\item [-o=False, -{-}overwrite=False]  
overwrite existing output files
\item [-q, -{-}qsub]  
a group ID under which to submit jobs to the queue
\item [-{-}name]  
a job name (only used with --qsub; if not set, --out basename will be used
\item [-d=/usr3/bustaff/koesterr/UGA/doc, -{-}directory=/usr3/bustaff/koesterr/UGA/doc]  
an output directory path
\item [-{-}cpus=1]  
number of cpus (limited module availability)
\item [-{-}mem=3]  
amount of ram memory to request for queued job (in gigabytes)
\item [-r, -{-}region]  
a region specified in tabix format (ie. 1:10583-1010582).
\item [-{-}region-list]  
a filename for a list of tabix format regions
\item [-s=False, -{-}split=False]  
split region list entirely into separate jobs (requires --region-list)
\item [-n, -{-}split-n]  
split region list into n separate jobs (requires --region-list)
\item [-{-}chr]  
chromosome number (1-26)
\item [-{-}split-chr=False]  
split by chromosome (will generate up to 26 separate jobs depending on chromosome coverage)
\item [-j, -{-}job]  
run a particular job number (requires --split-n)
\item [-{-}job-list]  
a filename for a list of job numbers (requires --split-n)
\item [-c, -{-}cfg]  
a configuration file name
\item [-v, -{-}vars]  
a declaration of the form A=B, C=D, E=F, ... to replace {[}A{]} with B, {[}C{]} with D, {[}E{]} with F, ... in any line of the cfg file
\end{optionlist}

\end{description}


\section{Annotation}
\label{annot::doc}\label{annot:annotation}\begin{alltt}
usage: uga annot {[}-h{]} {[}--version{]} {[}-o{]} {[}-q QSUB{]} {[}--name NAME{]} {[}-d DIRECTORY{]}
                 {[}--cpus CPUS{]} {[}--mem MEM{]}
                 {[}-r REGION \textbar{} --region-list REGION\_LIST{]}
                 {[}-s \textbar{} -n SPLIT\_N \textbar{} --chr CHR \textbar{} --split-chr{]}
                 {[}-j JOB \textbar{} --job-list JOB\_LIST{]}
\end{alltt}
\begin{description}
\item[{Options:}] \leavevmode\begin{optionlist}{3cm}
\item [-{-}version]  
show program's version number and exit
\item [-o=False, -{-}overwrite=False]  
overwrite existing output files
\item [-q, -{-}qsub]  
a group ID under which to submit jobs to the queue
\item [-{-}name]  
a job name (only used with --qsub; if not set, --out basename will be used
\item [-d=/usr3/bustaff/koesterr/UGA/doc, -{-}directory=/usr3/bustaff/koesterr/UGA/doc]  
an output directory path
\item [-{-}cpus=1]  
number of cpus (limited module availability)
\item [-{-}mem=3]  
amount of ram memory to request for queued job (in gigabytes)
\item [-r, -{-}region]  
a region specified in tabix format (ie. 1:10583-1010582).
\item [-{-}region-list]  
a filename for a list of tabix format regions
\item [-s=False, -{-}split=False]  
split region list entirely into separate jobs (requires --region-list)
\item [-n, -{-}split-n]  
split region list into n separate jobs (requires --region-list)
\item [-{-}chr]  
chromosome number (1-26)
\item [-{-}split-chr=False]  
split by chromosome (will generate up to 26 separate jobs depending on chromosome coverage)
\item [-j, -{-}job]  
run a particular job number (requires --split-n)
\item [-{-}job-list]  
a filename for a list of job numbers (requires --split-n)
\end{optionlist}

\end{description}


\section{Plots}
\label{plot:plots}\label{plot::doc}\begin{alltt}
usage: uga plot {[}-h{]} {[}--version{]} {[}-o{]} {[}-q QSUB{]} {[}--name NAME{]} {[}-d DIRECTORY{]}
                {[}--cpus CPUS{]} {[}--mem MEM{]}
                {[}-r REGION \textbar{} --region-list REGION\_LIST{]}
                {[}-s \textbar{} -n SPLIT\_N \textbar{} --chr CHR \textbar{} --split-chr{]}
                {[}-j JOB \textbar{} --job-list JOB\_LIST{]}
\end{alltt}
\begin{description}
\item[{Options:}] \leavevmode\begin{optionlist}{3cm}
\item [-{-}version]  
show program's version number and exit
\item [-o=False, -{-}overwrite=False]  
overwrite existing output files
\item [-q, -{-}qsub]  
a group ID under which to submit jobs to the queue
\item [-{-}name]  
a job name (only used with --qsub; if not set, --out basename will be used
\item [-d=/usr3/bustaff/koesterr/UGA/doc, -{-}directory=/usr3/bustaff/koesterr/UGA/doc]  
an output directory path
\item [-{-}cpus=1]  
number of cpus (limited module availability)
\item [-{-}mem=3]  
amount of ram memory to request for queued job (in gigabytes)
\item [-r, -{-}region]  
a region specified in tabix format (ie. 1:10583-1010582).
\item [-{-}region-list]  
a filename for a list of tabix format regions
\item [-s=False, -{-}split=False]  
split region list entirely into separate jobs (requires --region-list)
\item [-n, -{-}split-n]  
split region list into n separate jobs (requires --region-list)
\item [-{-}chr]  
chromosome number (1-26)
\item [-{-}split-chr=False]  
split by chromosome (will generate up to 26 separate jobs depending on chromosome coverage)
\item [-j, -{-}job]  
run a particular job number (requires --split-n)
\item [-{-}job-list]  
a filename for a list of job numbers (requires --split-n)
\end{optionlist}

\end{description}


\chapter{Index and Search}
\label{index:index-and-search}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}
